File: backend/app/schemas/dataset.py
from pydantic import BaseModel, Field


class CreateDatasetRequest(BaseModel):
    source: str
    new_name: str
    prefix: str
    splits_to_include: list[str] = ["train", "valid", "test"]
    # Maps source class id (as string, since JSON object keys are strings) -> desired output name.
    # Omitting a source class id drops every box of that class from the copied labels.
    # None (the default) keeps every class with its original name.
    class_filter: dict[str, str] | None = None


class MergeDatasetRequest(BaseModel):
    source: str
    destination: str
    prefix: str
    splits_to_include: list[str] = ["train", "valid", "test"]
    class_filter: dict[str, str] | None = None


class MergeSourceEntry(BaseModel):
    source: str
    prefix: str
    class_filter: dict[str, str] | None = None


class MergeMultiRequest(BaseModel):
    destination: str
    sources: list[MergeSourceEntry]
    splits_to_include: list[str] = ["train", "valid", "test"]


class ResplitRequest(BaseModel):
    train: float = Field(0.8, ge=0, le=1)
    valid: float = Field(0.2, ge=0, le=1)
    test: float = Field(0.0, ge=0, le=1)


class DatasetSummary(BaseModel):
    name: str
    splits: dict[str, int]
    total_images: int
    num_classes: int
    last_modified: str | None
    size_bytes: int


class DatasetDetail(BaseModel):
    name: str
    splits: dict[str, int]
    total_images: int
    classes: dict[int, str]
    last_modified: str | None
    size_bytes: int
    history: list[dict]

File: frontend/src/types/dataset.ts
export interface DatasetSummary {
  name: string;
  splits: Record<string, number>;
  total_images: number;
  num_classes: number;
  last_modified: string | null;
  size_bytes: number;
}

export interface DatasetDetail {
  name: string;
  splits: Record<string, number>;
  total_images: number;
  classes: Record<string, string>;
  last_modified: string | null;
  size_bytes: number;
  history: HistoryEvent[];
}

export interface HistoryEvent {
  event: string;
  timestamp: string;
  source_dataset?: string;
  prefix?: string;
  images_added?: number;
  images_skipped?: number;
  splits?: Record<string, number>;
  class_remap?: { old: Record<string, string>; new: Record<string, string> };
  dropped_classes?: Record<string, string> | null;
  errors?: { file: string; reason: string }[] | null;
  techniques?: Record<string, number>;
  dropped_boxes?: number;
}

export interface CreateDatasetRequest {
  source: string;
  new_name: string;
  prefix: string;
  splits_to_include: string[];
  class_filter?: Record<string, string>;
}

export interface MergeDatasetRequest {
  source: string;
  destination: string;
  prefix: string;
  splits_to_include: string[];
  class_filter?: Record<string, string>;
}

export interface MergeSourceEntry {
  source: string;
  prefix: string;
  class_filter?: Record<string, string>;
}

export interface MergeMultiRequest {
  destination: string;
  sources: MergeSourceEntry[];
  splits_to_include: string[];
}
